### ISSUE TITLE
Hi Ben, can PINNS be used for system identification, for example harmonic oscillator only mu parameter is inverted while other parameters are known.

### ISSUE THREAD

**Issue by Subodh213** [benmoseley/harmonic-oscillator-pinn#6]
In actual problem, I have only data can this be used to discover all the parameters?
I tried using k and mu as unknown parameters but solution dont converge.

**Comment by jeduapf** [comment]
Hello Subdh213, I'm currently working on a similar task, and it appears that there's a correlation between epoch size and learning rate that affects convergence, particularly for equations with components of different orders of magnitude. For instance, in the case of the harmonic oscillator example, the parameters k and μ differ by a factor of 100, while μ is already 10 times different from x(t). Although I'm not entirely certain, I speculate that increasing the number of epochs, adjusting the learning rate, and potentially modifying the neural network's structure could aid in achieving convergence.

I tried two different approaches while maintaining the lambda regularization as in the original code, 10**4 to 1 in favor of data points.

1. I changed the network to pinn = FCN(1, 1, 20, 5) and ran it for 1.000.001 iterations. The values for k and μ 
2. I kept the network as pinn = FCN(1, 1, 32, 3) and ran it for 500.001 iterations. The values for k and μ 

Both methods converged to a OK solution, but they did not yield accurate values for k and μ ( maybe with more iterations ). Furthermore, I find it concerning that for such a simple model, I needed to iterate up to 100001 times! I am also interested in optimizing these parameters in different PDEs, including those with forced solutions such as the Harmonic Oscillator with  a force over time. Does anyone have insights on how to achieve this?

**20/5 NN 1.000.001 iterations**


![20_5_k](https://github.com/benmoseley/harmonic-oscillator-pinn/assets/59791487/10e029ca-9330-4b5c-a98c-76a2babf93d0)
![20_5_learning_k_mu](https://github.com/benmoseley/harmonic-oscillator-pinn/assets/59791487/46ce7a90-8ff9-4317-8d5b-3afbdfcc03af)
![20_5_Losses](https://github.com/benmoseley/harmonic-oscillator-pinn/assets/59791487/dee41861-c3d6-441f-838f-dbbe641b2e70)
![20_5_mu](https://github.com/benmoseley/harmonic-oscillator-pinn/assets/59791487/e0c1bdfd-ffda-4139-9542-2f1e80946767)

**32/3 NN 500.001 iterations**

![32_3_k](https://github.com/benmoseley/harmonic-oscillator-pinn/assets/59791487/4c84f481-7c8c-41f0-a2a9-93054b2c238b)
![32_3_learning_k_mu](https://github.com/benmoseley/harmonic-oscillator-pinn/assets/59791487/00c0023d-aeb8-4c1f-88e6-00755860f1e9)
![32_3_Losses](https://github.com/benmoseley/harmonic-oscillator-pinn/assets/59791487/75599ae0-1400-4b14-85ca-cd33cab1af22)
![32_3_mu](https://github.com/benmoseley/harmonic-oscillator-pinn/assets/59791487/e9604198-4e80-4fdf-be45-bc2998e37ccc)

[copy_paste_python_code.txt](https://github.com/benmoseley/harmonic-oscillator-pinn/files/14959755/copy_paste_python_code.txt)

**Comment by benmoseley** [comment]
the solutions above look like they are trying to become discontinuous - this makes me think that there are not enough collocation points for the physics loss, have you tried increasing the number of collocation points?

**Comment by benmoseley** [comment]
also note that the Adam optimiser tries to take step sizes of the same magnitude as the learning rate. So if your learning rate is 1e-3, optimising k from 0 to 400 would take a minimum of 400,000 training steps. So I recommend normalising mu and k to the range [-1,1], or using different learning rates depending on their expected magnitude

**Comment by jeduapf** [comment]
Hello Ben, 

Thank you a lot for your reply, it really helped me think more about it. 

So I tried with more collocation points in the physics loss and changing the learning rate to lower rates to have a more smooth loss towards the minima. I also increased the ephocs and my results are bellow my remarks.

I just didn't understand how I can normalize the constants. I mean, if I have a PDE such as 
  -  1 d2y/dt2 + 10000 dy/dt + 20 y = 0 
  - Divide all by 20
  - 1/20 d2y/dt2 + 500 dy/dt + 1 y = 0
 
I know I can divide the whole equation by one of the constants to normalize one of them, but the proportion between the constants will always be the same no matter what I do to the equation. If I change them in different ways I would also change the dynamics of the system. So what do you mean by normalizing? Could you recommend any articles in the subject please?

Moreover, since the objective is to identify the system by finding its constants, it should be more interesting to add another term in the loss function of the physics loss related to the input signal for different signals, right? What are your insights about it?

From my tries I guess a smoother loss => smaller step (learning rate) and a higher regularization to the data points loss (lambda) influences a lot the learning. Is it right ?

1.  900 collocation points & 90 data points 

- Initial mu = -2 
- initial k = 437
- epochs = 500k
- learning rate = 10^-4
- data loss regularization, lambda = 10^4

( Data loss and physics loss graphs here are with inverted titles ) 

![g](https://github.com/benmoseley/harmonic-oscillator-pinn/assets/59791487/6aab7da2-7914-442f-b29b-d81530a39826)
![pinn_00499889](https://github.com/benmoseley/harmonic-oscillator-pinn/assets/59791487/6594a754-d8a3-48d5-a39c-ae8d90cbab96)
![newplot (2)](https://github.com/benmoseley/harmonic-oscillator-pinn/assets/59791487/9d4b52b2-146c-458f-b66d-64434db0051f)
![newplot (3)](https://github.com/benmoseley/harmonic-oscillator-pinn/assets/59791487/6e6b3b32-241f-44ce-898b-828e9caa2f46)



2.  300 collocation points & 60 data points

- Initial mu = 13 
- initial k = 417
- epochs = 100k
- learning rate = 3*10^-4
- data loss regularization, lambda = 10^5

![learning_k_mu](https://github.com/benmoseley/harmonic-oscillator-pinn/assets/59791487/41d1f3cb-c02a-4c96-b68c-a433bf674335)
![pinn_00099778](https://github.com/benmoseley/harmonic-oscillator-pinn/assets/59791487/f3d90885-fb1d-4eed-888e-2256cfd655db)
![newplot (4)](https://github.com/benmoseley/harmonic-oscillator-pinn/assets/59791487/ae521f01-6b83-40e3-8610-73f8813a612a)
![newplot (5)](https://github.com/benmoseley/harmonic-oscillator-pinn/assets/59791487/a891d6f7-9be8-46ed-aaa9-3c52472615cd)


CLOSER to the END training

![newplot (6)](https://github.com/benmoseley/harmonic-oscillator-pinn/assets/59791487/c06f9e91-ff04-40e3-9483-51c7ae56df4a)
